File: code/Python/912-Sort an Array.py
# Given an array of integers nums, sort the array in ascending order.
#
# Example 1:
# Input: nums = [5,2,3,1]
# Output: [1,2,3,5]
#
# Example 2:
# Input: nums = [5,1,1,2,0,0]
# Output: [0,0,1,1,2,5]
#
# Constraints:
# 1 <= nums.length <= 50000
# -50000 <= nums[i] <= 50000

class Solution(object):
    def selectionSort(self, nums):
        n = len(nums)
        for i in range(n):
            for j in range(i, n):
                if nums[i] > nums[j]:
                    nums[i], nums[j] = nums[j], nums[i]
        return nums

    def bubbleSort(self, nums):
        n = len(nums)
        for i in range(n):
            for j in range(1, n - i):
                if nums[j - 1] > nums[j]:
                    nums[j - 1], nums[j] = nums[j], nums[j - 1]
        return nums

    def insertionSort(self, nums):
        n = len(nums)
        for i in range(1, n):
            while i> 0 and nums[i - 1] > nums[i]:
                nums[i - 1], nums[i] = nums[i], nums[i - 1]
                i -= 1
        return nums


    def sortArray(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        # return self.selectionSort(nums)
        # return self.bubbleSort(nums)
        return self.insertionSort(nums)
#         nums.sort()


if __name__ == '__main__':
    s = Solution()
    nums = [5,1,1,2,0,0]
    print(s.sortArray(nums))

File: code/912-Sort an Array/912-Sort an Array.c
/*
LeetCode 912 - Sort an Array

解题步骤（自动整理）
1. 先把题目目标拆成可操作的子任务（比较/统计/构造等）
2. 选择合适的数据结构并按一次遍历或分治步骤实现核心逻辑
3. 补齐边界条件（空输入、单元素、重复元素等）并返回结果
*/
// Given an array of integers nums, sort the array in ascending order.



// Example 1:

// Input: nums = [5,2,3,1]
// Output: [1,2,3,5]
// Example 2:

// Input: nums = [5,1,1,2,0,0]
// Output: [0,0,1,1,2,5]


// Constraints:

// 1 <= nums.length <= 5 * 10^4
// -5 * 10^4 <= nums[i] <= 5 * 10^4



/**
 * Note: The returned array must be malloced, assume caller calls free().
 */

int cmp(int* a, int* b) {
    return *a - *b;
}

int* sortArray(int* nums, int numsSize, int* returnSize){
    *returnSize = numsSize;
    qsort(nums, numsSize, sizeof(int), cmp);
    return nums;
}

 void swap(int* a, int* b) {
    int tmp = *a;
    *a = *b;
    *b = tmp;
}

// 二元选择排序

void swap(int* a, int* b) {
    int tmp = *a;
    *a = *b;
    *b = tmp;
}

int* sortArray(int* nums, int numsSize, int* returnSize){
    *returnSize = numsSize;
    for (int i = 0; i < numsSize / 2; i ++) {
        int minIndex = i, maxIndex = i;
        for (int j = i + 1; j < numsSize - i; j ++) {
            if (nums[j] < nums[minIndex]) {
                minIndex = j;
            }
            if (nums[j] > nums[maxIndex]) {
                maxIndex = j;
            }
        }
        if (minIndex == maxIndex) {
            break;
        }
        swap(nums + i, nums + minIndex);
        if (maxIndex == i) {
            maxIndex = minIndex;
        }
        swap(nums + numsSize - 1, nums + maxIndex);
    }
    return nums;
}

// 普通选择排序
int* sortArray(int* nums, int numsSize, int* returnSize){
    *returnSize = numsSize;
    for (int i = 0; i < numsSize; i ++) {
        int minIndex = i;
        for (int j = i + 1; j < numsSize; j ++) {
            if (nums[j] < nums[minIndex]) {
                minIndex = j;
            }
        }
        swap(nums + i, nums + minIndex);
    }
    return nums;
}
